designs/NuGet-ClientCertificates.md

```diff
@@ -0,0 +1,114 @@
+## Client certificates in NuGet
+
+Implements: https://github.com/NuGet/Home/issues/5773
+* Status: **Purposed**
+* Author(s): [Shkolka Volodymyr](https://github.com/BlackGad)
+
+## Issue
+
+[5773](https://github.com/NuGet/Home/issues/5773) - NuGet cannot restore from HTTPS sources that require Client Certificates
+
+## Background
+
+Developer tools that supports [Client Certificate Authentication](https://blogs.msdn.microsoft.com/kaushal/2015/05/27/client-certificate-authentication-part-1/).
+* [NPM](https://docs.npmjs.com/misc/config#cert)
+* [Maven](https://maven.apache.org/guides/mini/guide-repository-ssl.html)
+* [Teamcity](https://www.jetbrains.com/help/teamcity/using-https-to-access-teamcity-server.html#UsingHTTPStoaccessTeamCityserver-ConfiguringJVMforauthenticationwithclientcertificate)
+
+It is necessary to add client certificate authentication feature support to NuGet.
+
+## Who are the customers
+
+All .NET Core customers.
+
+## Requirements
+
+* Ability to specify client certificates from common [certificate stores](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/working-with-certificates#certificate-stores)
+* Ability to specify client certificates from external files in standard [formats](https://en.wikipedia.org/wiki/X.509#Certificate_filename_extensions)
+* Ability to specify inline client certificates in Base64 encoded DER certificate format
+
+## Solution
+
+1) Find specified in NuGet configuration file certificates (common implementation for [all products](https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Core/NuGet.Configuration) )
+2) Apply them prior any web request to appropriate http handler (product specific)
+
+## NuGet configuration changes
+
+Added new configuration section **clientCertificates** which may have children of 2 types:
+
+**fromStorage** - item for certificate import from [certificate store](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/working-with-certificates#certificate-stores). Internally uses [X509Certificate2Collection.Find](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2collection.find?view=netframework-4.8#System_Security_Cryptography_X509Certificates_X509Certificate2Collection_Find_System_Security_Cryptography_X509Certificates_X509FindType_System_Object_System_Boolean_) method.
```

Comment:
I haven't played with certificates on .NET Core. Are these APIs available in .NET Core, or is it .NET Framework only? Do they work only on Windows, or also on Linux and Mac?

Reply:
Accordingly to [spec](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2collection?view=netframework-4.8) this functionality defined for NET Standard 2.0 so I believe it is cross platform.